### PR TITLE
Add `table_exists` method to Catalog

### LINKED ISSUE #507: Add `table_exists` method to the Catalog
### Feature Request / Improvement

Prolonging the discussion in https://github.com/apache/iceberg-python/issues/406

I think having this method on the `Catalog` would be a great addition. On top of this, we can also leverage the table-exists implementation of the REST catalog by overriding this method in the REST catalog.

### PR DESCRIPTION
**Pull request by anupam-saini**
Closes #507

### CODE REVIEW

File: pyiceberg/catalog/__init__.py [hunk 1]
```diff
@@ -646,6 +646,13 @@ def purge_table(self, identifier: Union[str, Identifier]) -> None:
         delete_files(io, prev_metadata_files, PREVIOUS_METADATA)
         delete_files(io, {table.metadata_location}, METADATA)
 
+    def table_exists(self, identifier: Union[str, Identifier]) -> bool:
```
Comment (syun64) [inline pyiceberg/catalog/__init__.py:649]: Hi @anupam-saini This looks great! In addition to this default behavior, I think there is a suggestion on #507 to override this method in [catalog/rest.py](https://github.com/apache/iceberg-python/blob/pyiceberg-0.6.x/pyiceberg/catalog/rest.py) to instead make a head request to the Rest Catalog at [/v1/{prefix}/namespaces/{namespace}/tables/{table}](https://github.com/apache/iceberg/blob/main/open-api/rest-catalog-open-api.yaml#L814) endpoint. Shall we incorporate that suggestion to use the available endpoint in the Rest Catalog?

**Comment by Fokko** [comment]
This looks good @anupam-saini, thanks for working on this. One more request, could you also update the docs? Otherwise people won't be able to find this :)

**Comment by anupam-saini** [comment]
> This looks good @anupam-saini, thanks for working on this. One more request, could you also update the docs? Otherwise people won't be able to find this :)

Good point :)

**Review by Fokko — APPROVED** [review]
Thanks @anupam-saini for working on this 🚀 and thank @syun64 for the review!

**Pull request merged**